src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,120 @@
+    {
+        /// <summary>
+        /// Information about rename document calls that allows them to be applied as individual actions.
+        /// 
+        /// Updates a solution with the actions determined at analysis time. To apply all actions
+        /// use <see cref="UpdateSolutionAsync(Solution, CancellationToken)"/>, or use a subset
+        /// of the actions by calling <see cref="UpdateSolutionAsync(Solution, ImmutableArray{RenameDocumentAction}, CancellationToken)"/>. 
+        /// Each action has a description of the changes that it will apply that can be presented to a user.
+        /// </summary>
+        public sealed class RenameDocumentActionSet
+        {
+            private readonly DocumentId _documentId;
+            private readonly string _documentName;
+            private readonly IReadOnlyList<string> _documentFolders;
+            private readonly ProjectId _projectId;
+
+            internal RenameDocumentActionSet(
+                ImmutableArray<RenameDocumentAction> actions,
+                ProjectId projectId,
+                DocumentId documentId,
+                string documentName,
+                IReadOnlyList<string> documentFolders)
+            {
+                ApplicableActions = actions;
+                _documentFolders = documentFolders;
+                _documentId = documentId;
+                _documentName = documentName;
+                _projectId = projectId;
+            }
+
+            /// <summary>
+            /// All applicable actions computed for the action
+            /// </summary>
+            public ImmutableArray<RenameDocumentAction> ApplicableActions { get; }
+
+            /// <summary>
+            /// Same as calling <see cref="UpdateSolutionAsync(Solution, ImmutableArray{RenameDocumentAction}, CancellationToken)"/> with 
+            /// <see cref="ApplicableActions"/> as the argument
+            /// </summary>
+            public Task<Solution> UpdateSolutionAsync(Solution solution, CancellationToken cancellationToken)
+                => UpdateSolutionAsync(solution, ApplicableActions, cancellationToken);
+
+            /// <summary>
+            /// Applies each <see cref="RenameDocumentAction"/> in order and returns the final solution. 
+            /// All actions must be contained in <see cref="ApplicableActions" />
+            /// </summary>
+            /// <remarks>
+            /// An empty action set is still allowed and will return a modified solution
+            /// that will update the document properties as appropriate.
+            /// </remarks>
+            public async Task<Solution> UpdateSolutionAsync(Solution solution, ImmutableArray<RenameDocumentAction> actions, CancellationToken cancellationToken)
+            {
+                if (actions.Any(a => !ApplicableActions.Contains(a)))
+                {
+                    throw new InvalidOperationException("Cannot apply action that is not in ApplicableActions");
```

Comment:
these needs to be loc'ed and needs to be an ArgumentException.  nothing prevents somone from calling multiple times, and then passing invalid data from one call into a later set.

Reply:
Good catch